pandas/__init__.py

```diff
@@ -319,11 +319,37 @@ def __getattr__(self, item):
             except AttributeError:
                 raise AttributeError(f"module datetime has no attribute {item}")
 
-    datetime = __Datetime().datetime
+    datetime = __Datetime()
 
-    class SparseArray:
-        pass
+    class __SparseArray(pandas.core.arrays.sparse.SparseArray):
```

Comment:
I am not sure we can do it this way for SparseArray, as that breaks isinstance checks

Reply:
To confirm, you mean something like

```python
isinstance(pd.array([1, 2, 3], dtype="Sparse"), pd.SparseArray)
```

will be false? Since it's an instance of the parent `pd.arrays.SparseArray`, not the child `pd.SparseArray`?

Reply:
Yes.

(in general, deprecating moving classes is annoying)

For `datetime` and `np` the same is true of course, but I think it is much less likely that someone will do that.

Reply:
This is again another issue with python 3.6 handling the deprecation of a module attribute.  That's why they implemented PEP 562 (https://www.python.org/dev/peps/pep-0562/) for python 3.7.  Making a reference using python 3.6 to `pd.SparseArray` (without a constructor) produce a warning would be quite involved (see the Rationale in the PEP discussion), and I don't think it is worth introducing that level of complexity to pandas to handle deprecation messages of a class for users of an older version of python.

It's why I suggested above to merge in this pull request, and we can leave this as an open issue.

Reply:
I know it's an issue with python 3.6, but that we still support 3.6 now is a reality we have to live with. 
So my suggestion would be to simply don't try to raise a warning in python 3.6, and just import the existing class in that case (by the time we are going to enforce this deprecation, we maybe won't support py3.6 anymore, so then that is less of an issue). Or simply don't deprecate pd.SparseArray at all.

Reply:
> Or simply don't deprecate pd.SparseArray at all.

That's probably my preference until we support just 3.7+.

Though @Dr-Irv this PR is also fixing other issues with the other deprecations on master? Taking a closer look now.

Reply:
@TomAugspurger Yes, there were other issues with deprecations on master that I addressed, especially in the testing code.